src/main/java/teammates/common/util/TimeHelper.java

```diff
@@ -156,8 +156,50 @@ public static Date getMsOffsetToCurrentTimeInUserTimeZone(int offset, double tim
     }
 
     public static Calendar convertToUserTimeZone(Calendar time, double timeZone) {
-        time.add(Calendar.MILLISECOND, (int) (60 * 60 * 1000 * timeZone));
-        return time; // for chaining
+        int serverOffset = time.getTimeZone().getRawOffset();
+        int sessionOffset = (int) (60 * 60 * 1000 * timeZone);
+        int offset = sessionOffset - serverOffset;
+        TimeZone sessionTimeZone = getTimeZone(timeZone);
+
+        // Check the timeZone if it is already set
+        if (time.getTimeZone().equals(sessionTimeZone)) {
+            return time;
+        } else {
+            time.setTimeZone(sessionTimeZone);
+            time.add(Calendar.MILLISECOND, offset);
+        }
+        return time;
+    }
+    
+    public static TimeZone getTimeZone(double timeZone) {
+        int hours = (int) timeZone;
+        int minutes = (int) (timeZone - hours) * 60;
+
+        String timeZoneId = "GMT"
+                + String.format("%s%02d%02d", hours > 0 ? "+" : "", hours, Math.abs(minutes));
+
+        return TimeZone.getTimeZone(timeZoneId);
+    }
+    
+    /**
+     * Formats the timestamp to a String with respecting the time zone
```

Comment:
`respective`?

As this is a public method of a utility class,
1. The header comment should define the precise behavior.
2. Recommended to add unit tests

Same for other new methods in this class.